src/lib/tls/tls_extensions_key_share.cpp

```diff
@@ -0,0 +1,553 @@
+      group == Group_Params::BRAINPOOL384R1 ||
+      group == Group_Params::BRAINPOOL512R1;
+   }
+
+[[maybe_unused]] constexpr bool is_dh(const Group_Params group)
+   {
+   return
+      group == Group_Params::FFDHE_2048 ||
+      group == Group_Params::FFDHE_3072 ||
+      group == Group_Params::FFDHE_4096 ||
+      group == Group_Params::FFDHE_6144 ||
+      group == Group_Params::FFDHE_8192;
+   }
+
+class Key_Share_Entry
+   {
+   public:
+      Key_Share_Entry(TLS_Data_Reader& reader)
+         {
+         // TODO check that the group actually exists before casting...
+         m_group = static_cast<Named_Group>(reader.get_uint16_t());
+         const auto key_exchange_length = reader.get_uint16_t();
+         m_key_exchange = reader.get_fixed<uint8_t>(key_exchange_length);
+         }
+
+      Key_Share_Entry(Named_Group group, std::vector<uint8_t> key_exchange)
+         : m_group(group)
+         , m_key_exchange(std::move(key_exchange))
+         {
+         if(m_key_exchange.empty())
+            {
+            throw Decoding_Error("Size of key_exchange in KeyShareEntry must be at least 1 byte.");
+            }
+         }
+
+      Key_Share_Entry(const TLS::Group_Params group, Callbacks& cb, RandomNumberGenerator& rng)
+         : m_group(group)
+         {
+         if(is_ecdh(group))
+            {
+            const EC_Group ec_group(cb.tls_decode_group_param(group));
+            auto skey = std::make_unique<ECDH_PrivateKey>(rng, ec_group);
+
+            // RFC 8446 Ch. 4.2.8.2
+            //
+            //   Note: Versions of TLS prior to 1.3 permitted point format
+            //   negotiation; TLS 1.3 removes this feature in favor of a single point
+            //   format for each curve.
+            //
+            // Hence, we neither need to take Policy::use_ecc_point_compression() nor
+            // ClientHello::prefers_compressed_ec_points() into account here.
+            m_key_exchange = skey->public_value(PointGFp::UNCOMPRESSED);
+            m_private_key = std::move(skey);
+            }
+         else if(is_dh(group))
+            {
+            // RFC 8446 Ch. 4.2.8.1
+            //
+            //   The opaque value contains the Diffie-Hellman
+            //   public value (Y = g^X mod p) for the specified group (see [RFC7919]
+            //   for group definitions) encoded as a big-endian integer and padded to
+            //   the left with zeros to the size of p in bytes.
+            auto skey = std::make_unique<DH_PrivateKey>(rng, DL_Group(cb.tls_decode_group_param(group)));
+            m_key_exchange = skey->public_value();
+            m_private_key = std::move(skey);
+#if defined(BOTAN_HAS_CURVE_25519)
+            }
+         else if(is_x25519(group))
+            {
+            auto skey = std::make_unique<X25519_PrivateKey>(rng);
+            m_key_exchange = skey->public_value();
+            m_private_key = std::move(skey);
+#endif
+            }
```

Comment:
```suggestion
            }
#if defined(BOTAN_HAS_CURVE_25519)
         else if(is_x25519(group))
            {
            auto skey = std::make_unique<X25519_PrivateKey>(rng);
            m_key_exchange = skey->public_value();
            m_private_key = std::move(skey);
            }
#endif
```
This is more readable